proposals/0000-ghc-module-naming.rst

```diff
@@ -0,0 +1,139 @@
+
+  * It is user-facing (user are encouraged to depend on it), but its API is less stable than ``base``.
+
+* ``ghc-prim, ghc-internals`` (and perhaps others): define functions and data types used internally by GHC to support the API of ``base`` and ``ghc-experimental``.
+
+  * These libraries come with no stability guarantees: they may change at short notice.
+
+In addition we already have:
+
+* ``ghc``: this library exposes GHC as a library, through the (currently ill-defined) GHC API.
+
+All these libraries follow the Haskell Package Versioning Policy (PVP).
+
+The question arises of *what module names should be used*. For example, suppose that all three exposed a module called ``Data.Tuple``.  In principle that would be fine -- GHC allows you
+to use the package name in the ``import`` statement, to disambiguate.  But it's *extremely* confusing.  This proposal articulates a set of conventions to
+help us design module names.
+
+The proposal
+============
+
+This proposal is split into four sub-proposals for easier discussion.  Each sub-proposal builds on the
+earlier ones -- they are increments, not alternatives.
+
+The goals of this proposal are deliberately limited to establish naming conventions.  We do not propose
+any changes to ``ghc`` or to ``cabal``.
+
+Proposal 1
+-----------
+
+* Modules in ``base``, ``ghc-experimental``, ``ghc-prim``, ``ghc-internals`` etc should all have distinct names.
+
+That principle leads immediately to the question: what should those names be?  Hence proposal 2.
+
+Proposal 2
+-----------
+
+* Modules in GHC's internal libraries (``ghc-prim``, ``ghc-internals`` etc) should be of form ``GHC.*``.
+* Modules in ``ghc-experimental`` should be of form ``Experimental.*``.
+* Modules in ``base`` should not have either of these prefixes.
+
+So example we might have
+
+* ``GHC.Tuple`` in ``ghc-internals``,
+* ``Experimental.Tuple`` or ``Experimental.Data.Tuple`` in ``ghc-experimental``
+* ``Data.Tuple`` in ``base``
+
+Proposal 3
+-----------
+
+The current ``base`` API exposes many modules starting with ``GHC.*``, so the proposed conventions could only
+apply to *new* modules.
+
+* Over time, and only with the agreement and support of the Core Libraries Committee, we may remove some ``GHC.*`` modules
+  from ``base``, especially ones that are barely used, or are manifestly "internal" (i.e. part of the implementation
+  of other, more public functions).
+  Of course there would be a significant deprecation cycle, to allow client libraries to adapt.
+
+Proposal 3 only expresses a direction of travel.  We will have to see what the CLC's attitude is,
+and what the Haskell community thinks.  Anything that disturbs the API of base needs to be considered
+rather carefully.
+
+
+Proposal 4
+------------
+
+* The public API of package ``ghc`` (GHC as a library) should have modules of form ``GhcAPI.*``.
```

Comment:
Is there some particular reason to use a "merged prefix"(`GhcAPI.*`) instead of using the module structure to express the relationship (`GHC.API.*` - this is the `API` of `GHC`)? The latter feels more natural.